index.js

```diff
@@ -0,0 +1,269 @@
+          })
+          .catch((err) => {
+            this.log.error('[Eriksen] Problem with reprocessing', err);
+          });
+      } else {
+        schemas[message.model].schema.queueInProgress = message.queue;
+      }
+    }
+  }
+
+  drainQueue(model) {
+    let queueKey = schemas[model].schema.config.queueKey
+      , modelCallPromises = [];
+
+    return new Promise((resolve, reject) => {
+      this.getAllQueued(model, queueKey, modelCallPromises).then((allModelCalls) => {
+        return Promise.all(allModelCalls);
+      })
+      .then((resolutions) => {
+        resolve(resolutions.length);
+      })
+      .catch((err) => {
+        reject(err);
+      });
+    });
+  }
+
+  getAllQueued(model, queueKey, modelCallPromises) {
+    return this.getQueuedMessage(queueKey)
+      .then((message) => {
+        if (message === null) {
+          return modelCallPromises;
+        }
+
+        message = JSON.parse(message);
+
+        // For each queued command, apply the command to the relevant function with arguments
+        schemas[model].models.forEach((backend) => {
+          if (schemas[model].isSecondary(backend)) {
+            let functionPromise = backend.model[message.func].apply(backend.model, message.args);
+
+            modelCallPromises.push(
+              Promise.resolve(functionPromise)
+                .catch((err) => {
+                  schemas[model].schema.config.logger.error('[Eriksen] Error during reprocessing:', err);
+                  return true;
+                })
+            );
+          }
+        });
+
+        return this.getAllQueued(model, queueKey, modelCallPromises);
+      });
+  }
+
+  getQueuedMessage(queueKey) {
+    return new Promise((resolve, reject) => {
+      this.queue.lpop(queueKey, (err, message) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(message);
+        }
+      });
+    });
+  }
+
+  setQueueState(shouldQueue, messageKey) {
+    this.queue.publish(this.config.messageKey, JSON.stringify({ queue: shouldQueue, model: messageKey }));
+  }
+}
+
+class ModelProxy {
+  constructor(conf) {
+    this.name = conf.name;
+    this.models = new Map();
+    this.schema = conf.schema;
+    this.config = conf.config;
+    this._cache = new Map();
+  }
+
+  addModel(modelName, model) {
+    model.name = modelName;
+
+    this.models.set(modelName, model);
+
+    _.forOwn(model.model, (method, name) => {
+      if (_.isFunction(method)) {
+        this.schema[name] = this.proxify(name);
+      }
+    });
+  }
+
+  isPrimary(model) {
+    return this.schema.config.primary === model.name;
+  }
+
+  isSecondary(model) {
+    return this.schema.config.primary !== model.name;
+  }
+
+  proxify(name) {
+    // Can't use an ES6 shorthand here, need to be able to get arguments of
+    // called function (not inherited from the proxy's this value)
+    let self = this;
+
+    return function() {
```

Comment:
So each time you add a model it re-creates this function on the proxy instance, is that right?